lib/bundler/cli/outdated.rb

```diff
@@ -47,31 +50,41 @@ def run
       outdated_gems_list = []
 
       # Loop through the current specs
-      gemfile_specs, dependency_specs = current_specs.partition {|spec| current_dependencies.key? spec.name }
-      [gemfile_specs.sort_by(&:name), dependency_specs.sort_by(&:name)].flatten.each do |current_spec|
+      gemfile_specs, dependency_specs = current_specs.partition do |spec|
+        current_dependencies.key? spec.name
+      end
+
+      [gemfile_specs, dependency_specs].flatten.sort_by(&:name).each do |current_spec|
         next if !gems.empty? && !gems.include?(current_spec.name)
 
         dependency = current_dependencies[current_spec.name]
+        next if dependency.nil?
 
-        if strict
-          active_spec = definition.specs.detect {|spec| spec.name == current_spec.name && spec.platform == current_spec.platform }
+        spec_groups = if strict
+          [definition.specs.detect {|spec| spec.name == current_spec.name && spec.platform == current_spec.platform }]
         else
-          active_specs = definition.index[current_spec.name].select {|spec| spec.platform == current_spec.platform }.sort_by(&:version)
-          if !current_spec.version.prerelease? && !options[:pre] && active_specs.size > 1
-            active_spec = active_specs.delete_if {|b| b.respond_to?(:version) && b.version.prerelease? }
+          specs = definition.index[current_spec.name].select {|spec| spec.platform == current_spec.platform }.sort_by(&:version)
+          if !current_spec.version.prerelease? && !options[:pre] && specs.size > 1
+            specs.delete_if {|b| b.respond_to?(:version) && b.version.prerelease? }
+          else
+            specs
           end
-          active_spec = active_specs.last
         end
 
-        if options["filter-major"] || options["filter-minor"] || options["filter-patch"]
+        active_spec = Bundler.definition.gem_version_promoter.sort_versions(dependency, spec_groups).last
```

Comment:
@chrismo In master code is not calling to `gem_version_promoter.sort_versions`. I think it's a good idea keep calling to `gem_version_promoter`.

I created another PR https://github.com/bundler/bundler/pull/5114 for clean code stuff (I will remove the refactors from this PR).